### PR TITLE
transformations: First implementation of stencil-bufferize

### PR DESCRIPTION
**Pull request by PapyChacal**
A first implementation of stencil bufferization, enough to demonstrate on the main tests.

It is not enough for all goals in scope, but figured it's big enough to merge a first step.

### CODE REVIEW

File: xdsl/transforms/stencil_bufferize.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+                    result_types=[field_from_temp(o.type)],
+                )
+                if isa(o.type, TempType[Attribute])
+                else o
+            )
+            for o in op.operands
+        ]
+
+        loads = [
+            LoadOp(operands=[d], result_types=[r.type]) for r, d in zip(op.res, dests)
```
Comment (n-io) [inline xdsl/transforms/stencil_bufferize.py:81]:
```suggestion
            LoadOp(operands=[d], result_types=[r.type]) for d, r in zip(dests, op.res)
```

File: xdsl/transforms/stencil_bufferize.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        # We are interested in folding buffers of loaded values
+        if not isinstance(load, LoadOp):
+            return
+
+        underlying = load.field
+
+        # TODO: propery analysis of effects in between
+        # For illustration, only fold a single use of the handle
+        # (Requires more boilerplate to analyse the whole live range otherwise)
+        uses = handle.uses.copy()
```
Comment (n-io) [inline xdsl/transforms/stencil_bufferize.py:151]:
```suggestion
        uses = op.res.uses.copy()
```

File: xdsl/transforms/stencil_bufferize.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+
+    If someone did, we should probably insert a copy there. Right now, we just bail out.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: BufferOp, rewriter: PatternRewriter):
+
+        handle = op.res
+        # If this is a value-semantic buffer, we can't fold it
+        if not isinstance(handle.type, FieldType):
```
Comment (n-io) [inline xdsl/transforms/stencil_bufferize.py:137]:
```suggestion
        if not isinstance(op.res.type, FieldType):
```

File: xdsl/transforms/stencil_bufferize.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+    The underlying buffer of values resulting from a load of a buffer is the buffer
+    loaded from. That is, if nobody annoyed the said buffer in-between.
+
+    If someone did, we should probably insert a copy there. Right now, we just bail out.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: BufferOp, rewriter: PatternRewriter):
+
+        handle = op.res
```
Comment (n-io) [inline xdsl/transforms/stencil_bufferize.py:135]:
```suggestion
```

File: xdsl/transforms/stencil_bufferize.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+        new_load = LoadOp.create(
+            operands=[op.field],
+            result_types=[r.type for r in load.results],
+            attributes=load.attributes.copy(),
+            properties=load.properties.copy(),
+        )
+
+        rewriter.replace_op(apply, new_apply)
+        rewriter.replace_op(load, new_load)
+        rewriter.erase_op(op)
```
Comment (PapyChacal) [inline xdsl/transforms/stencil_bufferize.py:245]: Question to my rested self

Why is the Apply considered here already? Can't it just be a LoadStoreFold ?

File: xdsl/transforms/stencil_bufferize.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+        underlying = load.field
+
+        # TODO: propery analysis of effects in between
+        # For illustration, only fold a single use of the handle
+        # (Requires more boilerplate to analyse the whole live range otherwise)
+        uses = handle.uses.copy()
+        if len(uses) > 1:
+            return
+        user = uses.pop().operation
```
Comment (n-io) [inline xdsl/transforms/stencil_bufferize.py:154]: it's probably not safe to call `pop()` if there is a control-flow path in which you return without rewriting the op.

Comment (PapyChacal) [reply xdsl/transforms/stencil_bufferize.py:154]: Hence the `.copy()` above :wink: 

There must be more poetic ways, propositions welcome!

File: xdsl/transforms/stencil_bufferize.py (hunk 5 above)
Comment (PapyChacal) [reply xdsl/transforms/stencil_bufferize.py:244]: There are ways but it requires finer analysis, on shapes as on more precise side effects.
So, ditching for now, as the first goal is only to get it satisfyingly working on existing stencil workloads.

**Review by georgebisbas — APPROVED** [review]
Thanks!

**Pull request merged**